BedrockServer.cpp

```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
                         if (hasSocket) {
                             // Set the destructor callback for when the command finishes.
                             command->destructionCallback = &callback;
-
-                            // And lock the mutex so the command can't complete until we are in `wait` below.
-                            lock.lock();
                         }
 
-                        // Now we'll queue this command in one of two queues.
+                        // Now we queue or run this command.
                         auto _syncNodeCopy = atomic_load(&_syncNode);
                         if (_syncNodeCopy && _syncNodeCopy->getState() == SQLiteNode::STANDINGDOWN) {
                             _standDownQueue.push(move(command));
                         } else {
-                            SINFO("Queuing new '" << command->request.methodLine << "' command from local client, with "
-                                  << _commandQueue.size() << " commands already queued.");
-                            _commandQueue.push(move(command));
+                            // If this command is scheduled in the future, we can't just run it, we need to enqueue it to run at that point.
+                            // This functionality will go away as we remove the queues from bedrock, and so this can be removed at that time.
+                            if (command->request.calcU64("commandExecuteTime") > STimeNow()) {
+                                // These are implicitly fire-and-forget commands and a response will already have been sent in `buildCommandFromRequest`.
+                                _commandQueue.push(move(command));
+                                continue;
+                            } else {
+                                SINFO("Running new '" << command->request.methodLine << "' command from local client, with " << _commandQueue.size() << " commands already queued.");
+                                runCommand(move(command));
+                            }
                         }
 
-                        // Now that the command is queued, we wait for it to complete (if it's has a socket). When it's
-                        // destructionCallback fires, this will stop blocking and we can move on to the next request.
-                        // NOTE: This doesn't correctly handle spurious wakeups.
-                        if (hasSocket) {
-                            cv.wait(lock);
+                        // Now that the command is queued, we wait for it to complete (if it's has a socket, and hasn't finished by the time we get to this point).
+                        // When this happens, destructionCallback fires, sets `finished` to true, and we can move on to the next request.
+                        unique_lock<mutex> lock(m);
+                        if (!finished && hasSocket) {
+                            cv.wait(lock, [&]{return finished.load();});
                         }
```

Comment:
I see that `// NOTE: This doesn't correctly handle spurious wakeups.` was removed, so I assume that's what `finished` is fore. Can you walk me through the logic here?